src/core/annotation.js

```diff
@@ -877,17 +888,315 @@ class WidgetAnnotation extends Annotation {
+        return evaluator
+          .getOperatorList({
+            stream,
+            task,
+            resources: this.fieldResources,
+            operatorList,
+          })
+          .then(function () {
+            operatorList.addOp(OPS.endAnnotation, []);
+            return operatorList;
+          });
+      }
+    );
+  }
+
+  async getAppearance(evaluator, task, annotationStorage) {
+    // If it's a password textfield then no rendering to avoid to leak it.
+    // see 12.7.4.3, table 228
+    if (!annotationStorage || this.data.isPassword) {
+      return null;
+    }
+    const value = annotationStorage[this.data.id] || "";
+    if (value === "") {
+      return null;
+    }
+
+    // Magic value
+    const defaultPadding = 2;
+
+    // Default horizontal padding: can we have an heuristic to guess it?
+    const hPadding = defaultPadding;
+    const totalHeight = this.data.rect[3] - this.data.rect[1];
+    const totalWidth = this.data.rect[2] - this.data.rect[0];
+
+    const fontInfo = await this.getFontData(evaluator, task);
+    const [font, fontName] = fontInfo;
+    let [, , fontSize] = fontInfo;
+
+    fontSize = this.computeAutoSizedFont(
+      fontName,
+      fontSize,
+      totalHeight,
+      2 * defaultPadding
+    );
+
+    let descent = font.descent;
+    if (isNaN(descent)) {
+      descent = 0;
+    }
+
+    const vPadding = defaultPadding + Math.abs(descent) * fontSize;
+    const defaultAppearance = this.data.defaultAppearance;
+
+    if (this.data.comb) {
+      const combWidth = (totalWidth / this.data.maxLen).toFixed(2);
+      let buf = `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 ${hPadding} ${vPadding} Tm`;
+      let first = true;
+      for (const character of value) {
+        if (first) {
+          buf += ` (${character}) Tj`;
+          first = false;
+        } else {
+          buf += ` ${combWidth} 0 Td (${character}) Tj`;
+        }
+      }
+      buf += " ET Q EMC";
+      return buf;
+    }
+
+    if (this.data.multiLine) {
+      const renderedText = this.handleMultiline(
+        value,
+        font,
+        fontSize,
+        totalWidth,
+        alignment,
+        hPadding,
+        vPadding
+      );
+      return `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 0 ${totalHeight} Tm ${renderedText} ET Q EMC`;
+    }
+
+    const alignment = this.data.textAlignment;
+    if (alignment === 0 || alignment > 2) {
+      // Left alignment: nothing to do
+      return `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 ${hPadding} ${vPadding} Tm (${value}) Tj ET Q EMC`;
+    }
+
+    const renderedText = this.renderPDFText(
+      value,
+      font,
+      fontSize,
+      totalWidth,
+      alignment,
+      hPadding,
+      vPadding
+    );
+    return `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 0 0 Tm ${renderedText} ET Q EMC`;
+  }
+
+  async getFontData(evaluator, task) {
+    const operatorList = new OperatorList();
+    const initialState = new EvalState();
```

Comment:
Really sorry, but I'm going to have to come back to my previous comments about how using an `initialState` in this way is (for a lack of a better word) "abusing" the `getOperatorList` method in ways that were never intended and why it unfortunately isn't *guaranteed* to always work.

It would actually be fairly easy to create a *theoretically valid* PDF document, where the `intentState` would still be *empty* even after parsing of the `defaultAppearance` has completed. Basically, simply wrapping the `defaultAppearance` in save/restore operators would break the code below which relies on `initialState` to contain the desired properties (note how the `StateManager`, by necessity, deals with save/restore ops).

Reply:
So, what could be a good way to get font and font size ?
Or maybe I can use a special state which clones to itself and then get the last used font from it, wdyt ?

Reply:
> So, what could be a good way to get font and font size ?

I honestly don't know, off the top of my head, since I don't believe that the `getOperatorList` functionality was really created for the particular use-cases you have mind here.

> Or maybe I can use a special state which clones to itself and then get the last used font from it, wdyt ?

Assuming it won't negatively impact performance of the "normal" PDF document parsing, that could possibly work. (Although it still, to me, feels somewhat unfortunate that we're having to resort to these sort of tricks :-)  

---

Of course, you might also want to see check if @brendandahl has any better ideas (than me).